Wrappers/Python/ccpi/optimisation/functions/FGP_TV.py

```diff
@@ -0,0 +1,275 @@
+# -*- coding: utf-8 -*-
+#  CCP in Tomographic Imaging (CCPi) Core Imaging Library (CIL).
+
+#   Copyright 2017-2020 UKRI-STFC
+#   Copyright 2017-2020 University of Manchester
+
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+
+#   http://www.apache.org/licenses/LICENSE-2.0
+
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from ccpi.optimisation.functions import Function, MixedL21Norm
+from ccpi.optimisation.operators import Gradient
+import numpy 
+import functools
+
+class FGP_TV(Function):
+    
+    r'''Fast Gradient Projection algorithm for Total Variation(TV) Denoising (ROF problem)
+    
+    .. math::  \min_{x} \alpha TV(x) + \frac{1}{2}||x-b||^{2}_{2}
+                
+    Parameters:
+      
+      :param domain: Domain of the reconstruction
+      :param regularising_parameter: TV regularising parameter 
+      :param inner_iterations: Iterations of FGP algorithm
+      :param tolerance: Stopping criterion (Default=1e-4)
+      :param nonnegativity: 
+      
+    Reference:
+      
+        Beck, A. and Teboulle, M., 2009. Fast Gradient-Based Algorithms for Constrained
+        Total Variation Image Denoising and Deblurring Problems 
+        IEEE TRANSACTIONS ON IMAGE PROCESSING, VOL. 18, NO. 11, NOVEMBER 2009.
+        
+    '''    
+    
+    
+    def __init__(self, domain, regularising_parameter, inner_iterations, tolerance, nonnegativity):
+        
+        self.regularising_parameter = regularising_parameter
+        self.inner_iterations = inner_iterations
+        self.tolerance = tolerance
+        self.nonnegativity = nonnegativity                
+        self.gradient = Gradient(domain)
+        self.TV = regularising_parameter * MixedL21Norm()                 
+            
+    def __call__(self,x):
+        
+        # evaluate objective function of TV gradient
+        return self.TV(self.gradient.direct(x))
+    
+    
+    def projection_C(self, x, out=None):   
+
+        # This can be replaced by IndicatorBox(lower=0), but is less effient.
```

Comment:
How much more efficient without using IndicatorBox?  There is the usual trade-off "speed vs generality" here: 

Benefit of using the current hardcoded implementation is: 
1. faster (slightly or a lot?).  

Benefit of using IndicatorBox: 
1. No repeated code (if IndicatorBox is updated, perhaps a bug fixed, it is automatically updated everywhere IndicatorBox is used, whereas if code is repeated, one needs to remember to update that code too). 
2. more general functionality for example can also apply any lower bound and any upper bound, so we can solve a larger class of problems.